src/audio/dai.c

```diff
@@ -695,8 +695,21 @@ static int dai_comp_trigger_internal(struct comp_dev *dev, int cmd)
 		COMPILER_FALLTHROUGH;
 	case COMP_TRIGGER_STOP:
 		comp_dbg(dev, "dai_comp_trigger_internal(), STOP");
+/*
+ * Some platforms cannot just simple disable
+ * DMA channel during the transfer,
+ * because it will hang the whole DMA controller.
+ * Therefore, stop the DMA first and let the DAI
+ * drain the FIFO in order to stop the channel
+ * as soon as possible.
+ */
+#if CONFIG_DMA_SUSPEND_DRAIN
 		ret = dma_stop(dd->chan);
 		dai_trigger(dd->dai, cmd, dev->direction);
+#else
+		dai_trigger(dd->dai, cmd, dev->direction);
+		ret = dma_stop(dd->chan);
+#endif
 		break;
 	case COMP_TRIGGER_PAUSE:
```

Comment:
@iuliana-prodan shouldn't we do the same thing at start? I mean the order of dai/dma start/stop should be in reverse one to each other.

Reply:
At first, I thought the same thing.
But I've looked at the explanation for DMA_SUSPEND_DRAIN, from [Kconfig](https://github.com/thesofproject/sof/blob/master/Kconfig#L70) and this seems to happen only on stop.
Also, in the Intel's DMA driver this is only checked on [dw_dma_stop()](https://github.com/thesofproject/sof/blob/master/src/drivers/dw/dma.c#L389).

Reply:
Ok, I see. So, I'm fine with patches 1 and 2. I need to better review 3) and I think it is fine. I have my doubts about 4.